pkg/providers/instancetype/instancetype.go

```diff
@@ -152,7 +161,12 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1.EC2NodeClass)
 			}
 		})
 
-		it := p.instanceTypesResolver.Resolve(ctx, i, zoneData, nodeClass)
+		var vmMemoryOverhead int64
+		if cached, ok := p.vmMemoryOverheadCache.Get(fmt.Sprintf("%s-%016x", *i.InstanceType, amiHash)); ok {
+			vmMemoryOverhead = cached.(int64)
+		}
+
+		it := p.instanceTypesResolver.Resolve(ctx, i, vmMemoryOverhead, zoneData, nodeClass)
```

Comment:
Making these updates before we insert into the cache isn't ideal. This means we won't use any new discovered overheads until the previous cache entry expires. Unfortunately, there's no trivial way around it without additional changes: modifying the instance types directly after retrieving them from the cache will result in concurrent read-writes, and copying the data will result in significant memory usage.

Some possible solutions:
- Wrap the instance types in a synchronized object so we can mutate the instance capacity after insertion into the cache
- Surface a method on the cloudprovider to get per instance-type capacity overrides that the scheduler of can take advantage of. 
- Accept the 5 minute delay 

Personally I'm leaning toward the second option, it seems like it should integrate nicely with https://github.com/kubernetes-sigs/karpenter/pull/1305 if we end up taking that feature in.

Reply:
As you noted the cache only lives for 5min and I thought that was an acceptable delay. You made a suggestion in another comment about tracking actual capacity. Instead of overhead of passing through the overhead value would it be acceptable to check the capacity cache (would rename from vmMemoryOverheadCache) for each instance-type after doing a lookup on instance-type cache?
```
if item, ok := p.instanceTypesCache.Get(key); ok {
	instanceTypes := append([]*cloudprovider.InstanceType{}, item.([]*cloudprovider.InstanceType)...)
	for _, instanceType := range instanceTypes {
		if cachedCapacity, found := p.vmMemoryOverheadCache.Get(fmt.Sprintf("%s-%016x", instanceType, amiHash)); found {
			instanceType.Capacity[corev1.ResourceMemory] = cachedCapacity.(resource.Quantity)
		}
	}
}
```
Do you have any thoughts of what an extension in the cloudprovider would look like? Probably something [in here](https://github.com/kubernetes-sigs/karpenter/blob/a860293883b67cbda0b8f7917130ccd6857b6cc3/pkg/controllers/provisioning/provisioner.go#L250-L271)
I think having the ability to override would be useful but I also think that cloudproviders should be returning valid information from the beginning rather than treating the override extension to arrive there.

NodeOverlays aren't dynamic though and is why this shouldn't _only_ be housed there in my opinion. Taking the example:
```
# Add memory overhead of 10Mi to all instances with 2Gi memory
# https://github.com/aws/karpenter-provider-aws/issues/5161
kind: NodeOverlay
metadata:
  name: discount
spec:
  selector:
    matchLabels:
      karpenter.k8s.aws/instance-memory: 2048
  overhead:
    memory: 10Mi
```
That would require a user to list out every possible instance-type and configure a per instance-type overhead. As I said having the ability to configure overrides will be useful, but I still think an internal controller is important for accurately tracking instance-type capacity.

Reply:
@jmdeal Rather than adding sync logic I've introduced `vmCapacityCacheSeqNum` to address the issue with waiting for cache to expire. I've also adjusted things to override capacity directly.